app/domain/authentication/strategy.rb

```diff
@@ -77,6 +79,22 @@ def conjur_token(input)
       raise e
     end
 
+    def login(input)
+      authenticator = authenticators[input.authenticator_name]
+
+      validate_authenticator_exists(input, authenticator)
+      validate_security(input)
+
+      key = authenticator.login(input)      
+      raise InvalidCredentials unless key
+
+      audit_success(input)
+      new_login(input, key)
+    rescue => err
+      audit_failure(input, err)
+      raise err
+    end
+
```

Comment:
Having this extra method on `Strategy` pushes me just to the edge (maybe over) of how far I like to stretch SRP.  If I rewriting this today I'd use two command objects, perhaps `ExchangeCredentialsForToken` (previously `Strategy`)  and `ExchangeCredentialsForApiKey`.

All of the shared logic would be pulled out into objects, which both of them would delegate to: eg, `ValidateSecurity`, `ValidateAuthenticatorExists`, etc....

Reply:
Tracking here: https://github.com/cyberark/conjur/issues/747